csharp/src/Apache.Arrow/Ipc/ArrowStreamReader.cs

```diff
@@ -14,282 +14,61 @@
 // limitations under the License.
 
 using System;
-using System.Buffers;
-using System.Collections.Generic;
-using System.Diagnostics;
 using System.IO;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Apache.Arrow.Ipc
 {
+    /// <summary>
+    /// Represents a reader that can read Arrow streams.
+    /// </summary>
     public class ArrowStreamReader : IArrowReader, IDisposable
     {
-        public Schema Schema { get; protected set; }
-        public Stream BaseStream { get; }
+        private protected readonly ArrowReaderImplementation _implementation;
 
-        protected ArrayPool<byte> Buffers { get; }
+        public Schema Schema => _implementation.Schema;
 
         public ArrowStreamReader(Stream stream)
         {
-            BaseStream = stream ?? throw new ArgumentNullException(nameof(stream));
-            Buffers = ArrayPool<byte>.Create();
-            Schema = null;
-        }
-
-        protected bool HasReadSchema => Schema != null;
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
 
-        public virtual async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationToken cancellationToken = default)
-        {
-            // TODO: Loop until a record batch is read.
-            cancellationToken.ThrowIfCancellationRequested();
-            return await ReadRecordBatchAsync(cancellationToken);
+            _implementation = new ArrowStreamReaderImplementation(stream);
```

Comment:
Should this type test the stream and use the previously shown specialized implementation for FileStream?

Reply:
No, I don't think so. My understanding of Arrow is that there are really 2 formats:

1. An Arrow "stream" - https://arrow.apache.org/docs/format/IPC.html#streaming-format
2. An Arrow "file" - https://arrow.apache.org/docs/format/IPC.html#file-format

You can see that the arrow file has a magic header and footer `ARROW1` and a `<FOOTER>` that contains the schema.  The Arrow "stream" format doesn't. So the user needs to tell the library which one they are dealing with. The way this works is the user either creates `new ArrowStreamReader` for (1) Arrow "stream" format and `new ArrowFileReader` for (2) Arrow file format.